pennylane/templates/subroutines/controlled_sequence.py

```diff
@@ -0,0 +1,127 @@
+
+    num_wires = AnyWires
+    # grad_method = None
+
+    def _flatten(self):
+        return (self.base,), (self.control,)
+
+    @classmethod
+    def _unflatten(cls, data, metadata):
+        return cls(data[0], control=metadata[0])
+
+    def __init__(self, base, control, id=None):
+
+        control_wires = Wires(control)
+
+        if len(Wires.shared_wires([base.wires, control_wires])) != 0:
+            raise ValueError("The control wires must be different from the base operation wires.")
+
+        self.hyperparameters["control_wires"] = control_wires
+        self.hyperparameters["base"] = base
+
+        self._name = "ControlledSequence"
+
+        super().__init__(base, id=id)
+
+    @property
+    def hash(self):
+        return hash(
+            (
+                str(self.name),
+                self.control,
+                self.base.hash,
+            )
+        )
+
+    @property
+    def control(self):
+        """The control wires for the sequence"""
+        return self.hyperparameters["control_wires"]
+
+    @property
+    def wires(self):
+        return self.control + self.base.wires
+
+    @property
+    def has_matrix(self):
+        return False
+
+    def __repr__(self):
+        return f"ControlledSequence({self.base}, control={list(self.control)})"
+
+    def map_wires(self, wire_map: dict):
+        # pylint:disable=protected-access
+        new_op = copy(self)
+        new_op.hyperparameters["base"] = self.base.map_wires(wire_map=wire_map)
+        new_op.hyperparameters["control_wires"] = Wires(
+            [wire_map.get(wire, wire) for wire in self.control]
+        )
+        return new_op
+
+    def decomposition(self):  # pylint:disable=arguments-differ
+        r"""Representation of the operator as a product of other operators.
+
+        .. math:: O = O_1 O_2 \dots O_n.
+
+        .. seealso:: :meth:`~.CtrlSequence.decomposition`.
+
+        Args:
+            base (Operator): the operator that acts as the base for the sequence
+            control_wires (Any or Iterable[Any]): the control wires for the sequence
+
+        Returns:
+            list[.Operator]: decomposition of the operator
+        """
+        # ToDo: docstring missing example
+        return self.compute_decomposition(self.base, self.control)
+
+    @staticmethod
+    def compute_decomposition(base, control_wires):  # pylint:disable=arguments-differ
+        r"""Representation of the operator as a product of other operators.
+
+        .. math:: O = O_1 O_2 \dots O_n.
+
+        .. seealso:: :meth:`~.CtrlSequence.decomposition`.
+
+        Args:
+            base (Operator): the operator that acts as the base for the sequence
+            control_wires (Any or Iterable[Any]): the control wires for the sequence
+
+        Returns:
+            list[.Operator]: decomposition of the operator
+        """
+        # ToDo: docstring missing example
+        ops = []
+        powers_of_two = 1 << np.arange(len(control_wires))
```

Comment:
no, will switch

Reply:
note - switched again to a slightly uglier implementation that doesn't use numpy, because somehow not using numpy to generate the powers makes the decomposition work 🤯